File: rust/proto/src/unmo.rs
extern crate rand;
use self::rand::Rng;

use responder::*;

enum ResponderType {
    What,
    Random,
}

pub struct Unmo {
    pub name: String,
    resp_what: WhatResponder,
    resp_random: RandomResponder,
    current: ResponderType,
}

impl Unmo {
    pub fn new(name: &str) -> Unmo {
        Unmo {
            name: name.to_string(),
            resp_what: WhatResponder::new("What"),
            resp_random: RandomResponder::new("Random"),
            current: ResponderType::Random,
        }
    }

    fn responder(&self) -> &Responder {
        match self.current {
            ResponderType::Random => &self.resp_random,
            ResponderType::What   => &self.resp_what,
        }
    }

    pub fn dialogue(&mut self, input: String) -> String {
        let n = rand::thread_rng().gen_range(0,2);
        match n {
            0 => self.current = ResponderType::Random,
            1 => self.current = ResponderType::What,
            _ => ()
        }
        self.responder().response(input)
    }

    pub fn responder_name(&self) -> String {
        self.responder().name()
    }
}

File: python/proto/unmo.py
from random import choice, randint
from responder import RandomResponder, WhatResponder
from dictionary import Dictionary

class Unmo:
    def __init__(self, name):
        self._name = name
        self._dic  = Dictionary()
        self._resp_random = RandomResponder('Random', self._dic)
        self._resp_what   = WhatResponder('What')

    def dialogue(self, line):
        r = randint(0, 100)
        if r < 10:
            self._responder = self._resp_what
        else:
            self._responder = self._resp_random

        return self._responder.response(line)

    @property
    def responder_name(self):
        return self._responder.name

    @property
    def name(self):
        return self._name
